File: lib/src/main/java/io/github/kory33/guardedqueries/core/utils/extensions/StringSetExtensions.java
package io.github.kory33.guardedqueries.core.utils.extensions;

import java.util.Collection;

public class StringSetExtensions {
    private StringSetExtensions() {
    }

    public static boolean isPrefixOfSome(final Collection<String> strings, String string) {
        for (final var s : strings) {
            if (s.startsWith(string)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Pick a string starting with {@code startingWith},
     * that is not a prefix of any string from {@code strings}.
     */
    public static String freshPrefix(final Collection<String> strings, final String startingWith) {
        long count = 0;
        while (true) {
            final var candidate = startingWith + Long.toHexString(count);
            if (!isPrefixOfSome(strings, candidate)) {
                return candidate;
            }
            count += 1;
        }
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/StringSetExtensions.scala
package io.github.kory33.guardedqueries.core.utils.extensions

import scala.annotation.tailrec

object StringSetExtensions {
  given Extensions: {} with
    extension (strings: Set[String])
      def someStartsWith(string: String): Boolean = {
        strings.exists(_.startsWith(string))
      }

      /**
       * Pick a string starting with {@code prefix}, that is not a prefix of any string from
       * {@code strings}.
       */
      def freshPrefixStartingWith(prefix: String): String = {
        @tailrec def searchForFreshPrefix(count: Long): String = {
          val candidate = prefix + count.toHexString
          if (!strings.someStartsWith(candidate)) candidate
          else searchForFreshPrefix(count + 1)
        }

        searchForFreshPrefix(0)
      }
}
